### ISSUE TITLE
Remove forked kubectl drain code in favor of direct module usage

### ISSUE THREAD

**Issue by detiber** [kubernetes-sigs/cluster-api#3970]
Previously we copied kubectl drain code into the third_party directory with the intention of eventually removing the copy and importing upstream directly (once the kubectl move to staging was complete). The kubectl code is now published at k8s.io/kubectl, however it appears that there have been changes made locally to the drain code we have copied.

These changes need to be re-aligned with upstream and we should be relying on an upstream import.

**Comment by neolit123** [comment]
> Previously we copied kubectl drain code into the third_party directory with the intention of eventually removing the copy and importing upstream directly (once the kubeadm move to staging was complete). The kubeadm code is now published at k8s.io/kubeadm, however it appears that there have been changes made locally to the drain code we have copied.

`s/kubeadm/kubectl/` :)

**Comment by detiber** [comment]
> `s/kubeadm/kubectl/` :)

:grimacing: fixed, thanks for catching that

**Comment by srm09** [comment]
From the code reading I did, one change that I can see is:

This particular method uses the https://github.com/kubernetes-sigs/cluster-api/blob/master/third_party/kubernetes-drain/cordon.go#L89-L94 supplied `context`, but the corresponding code in the upstream `kubectl` codebase does not use the supplied context. I am unsure how big of a deal that is, but it is different from our current functionality.

**Comment by detiber** [comment]
@srm09 indeed, it definitely complicates the re-alignment. I'm wondering if we also need an automated PR check that verifies that anything updated in `third_party` is done so as part of a sync from upstream vs local changes?

**Comment by ncdc** [comment]
As part of the upgrade to controller-runtime 0.7 (#3735), we modified the third party drain code to take in contexts. If we want to keep that modification, then I think we'll need to manually reconcile when updating our copy of the drain code.

Should we also submit PRs to k/k to update the drain code to take in contexts, assuming that's feasible? If we can do that, then we should be able to vendor the kubectl drain code properly, instead of having a copy of it (once we can upgrade to the newer kubectl tag).

**Comment by srm09** [comment]
I will open an issue in the k/k repo around this, and start a Slack thread about the same and see what folks have to say about that.

**Comment by srm09** [comment]
Opened a draft PR against `kubectl` https://github.com/kubernetes/kubernetes/pull/97078

**Comment by srm09** [comment]
kubernetes/kubernetes#97078 got merged, but these changes will be landing in 1.21
Created a backport PR kubernetes/kubernetes#98375 for the 1.20 release stream.

**Comment by vincepri** [comment]
The linked PR has been merged

**Comment by detiber** [comment]
@vincepri If all of the changes landed upstream, and we are pulling in k8s-1.21, can we remove the forked code from `third_party` and import the kubectl drain code directly now?

**Comment by vincepri** [comment]
Yes, that would be probably best. 

/reopen
/retitle Remove forked kubectl drain code in favor of direct module usage
/help

**Comment by vincepri** [comment]
@srm09 Can you follow-up upstream with the rest of the changes needeD?

**Comment by srm09** [comment]
I have the changes made locally, will post a PR in a day or two. Will update the issue accordingly.

**Comment by srm09** [comment]
Merged https://github.com/kubernetes/kubernetes/pull/100439 to `main` branch for k8s repo. Marked for v1.22 milestone.

**Comment by srm09** [comment]
/remove-lifecycle stale
Waiting for 1.22 release.
